File: walking_robot_sim/src/quadropted_controller_cpp/src/odometry/odometry_update.cpp
#include <cmath>

#include "quadropted_controller_cpp/odometry/odometry.hpp"

namespace quadropted {

double normalize_angle(double angle) {
    return std::atan2(std::sin(angle), std::cos(angle));
}

void update_odometry(OdometryState& state, double dt, double contact_count_coeff) {
    if (dt <= 0.0) return;

    double delta_x_total = 0.0;
    double delta_y_total = 0.0;
    double contact_sum = 0.0;
    int actual_contacts = 0;

    for (int i = 0; i < 4; ++i) {
        if (state.foot_contacts[i]) {
            double foot_rel_x = state.foot_positions[i].x();
            double foot_rel_y = state.foot_positions[i].y();

            if (state.prev_foot_positions[i].has_value()) {
                double delta_x = foot_rel_x - state.prev_foot_positions[i]->x();
                double delta_y = foot_rel_y - state.prev_foot_positions[i]->y();

                delta_x_total += delta_x;
                delta_y_total += -delta_y;
                contact_sum += contact_count_coeff;
                actual_contacts++;
            }

            state.prev_foot_positions[i] = Eigen::Vector2d(foot_rel_x, foot_rel_y);
        }
    }

    double avg_delta_x, avg_delta_y;

    if (contact_sum > 0) {
        avg_delta_x = delta_x_total / contact_sum;
        avg_delta_y = delta_y_total / contact_sum;
    } else {
        avg_delta_x = state.linear_velocity_x * dt;
        avg_delta_y = state.linear_velocity_y * dt;
    }

    state.append_delta(avg_delta_x, avg_delta_y);
    auto avg = state.average_delta();

    double cos_theta = std::cos(state.theta);
    double sin_theta = std::sin(state.theta);
    state.x += avg.first * cos_theta - avg.second * sin_theta;
    state.y += avg.first * sin_theta + avg.second * cos_theta;
}

}  // namespace quadropted

File: walking_robot_sim/src/quadropted_controller/scripts/QuadrupedOdometry/odometry_update.py
#!/usr/bin/env python3
"""
Обновление одометрии — чистая логика (без ROS зависимостей).
Вынесено из QuadrupedOdometryNode.py при декомпозиции.
"""

import math


def normalize_angle(angle):
    """Нормализовать угол до [-pi, pi]."""
    return math.atan2(math.sin(angle), math.cos(angle))


def update_odometry(state, dt, contact_count_coeff=0.65):
    """
    Обновить позицию одометрии на основе контактов лап или скоростей.

    Использует state.theta (текущий угол из IMU) для поворота смещений.
    НЕ изменяет state.theta — это делается отдельно в imu_callback.

    :param state: OdometryState
    :param dt: Время с последнего обновления (секунды)
    :param contact_count_coeff: Коэффициент вклада каждой лапы (по умолч. 0.65)
    :return: None (состояние обновляется inplace)
    """
    if dt <= 0.0:
        return

    delta_x_total, delta_y_total = 0.0, 0.0
    contact_count = 0

    for i in range(4):
        if state.foot_contacts[i]:
            foot_rel_x = state.foot_positions[i][0]
            foot_rel_y = state.foot_positions[i][1]

            if state.prev_foot_positions[i] is not None:
                delta_x = foot_rel_x - state.prev_foot_positions[i][0]
                delta_y = foot_rel_y - state.prev_foot_positions[i][1]

                delta_x_total += delta_x
                delta_y_total += -delta_y
                contact_count += contact_count_coeff

            state.prev_foot_positions[i] = (foot_rel_x, foot_rel_y)

    if contact_count > 0:
        delta_x = delta_x_total / contact_count
        delta_y = delta_y_total / contact_count

        state.delta_x_queue.append(delta_x)
        state.delta_y_queue.append(delta_y)

        avg_delta_x = sum(state.delta_x_queue) / len(state.delta_x_queue)
        avg_delta_y = sum(state.delta_y_queue) / len(state.delta_y_queue)

        state.x += (avg_delta_x * math.cos(state.theta) - avg_delta_y * math.sin(state.theta))
        state.y += (avg_delta_x * math.sin(state.theta) + avg_delta_y * math.cos(state.theta))
    else:
        # Обновление на основе команд скорости
        delta_x = state.linear_velocity_x * dt
        delta_y = state.linear_velocity_y * dt

        state.delta_x_queue.append(delta_x)
        state.delta_y_queue.append(delta_y)

        avg_delta_x = sum(state.delta_x_queue) / len(state.delta_x_queue)
        avg_delta_y = sum(state.delta_y_queue) / len(state.delta_y_queue)

        state.x += (avg_delta_x * math.cos(state.theta) - avg_delta_y * math.sin(state.theta))
        state.y += (avg_delta_x * math.sin(state.theta) + avg_delta_y * math.cos(state.theta))
